### PR TITLE
feat(mpc-contract): cancel_node_migration() function in contract

### LINKED ISSUE #3774: No way to cancel a started node migration, contrary to what the migration doc claims
### Background

`start_node_migration` stores a `DestinationNodeInfo` record that never expires and cannot be withdrawn. The only removal paths are a successful `conclude_node_migration`, or the orphan cleanup after resharing — which only removes records of accounts that left the participant set. An operator can overwrite the pending record via another `start_node_migration` call (revoking the previous destination key's ability to conclude), but there is no way back to "no migration pending": there is always one live record, concludable at any later time by whoever holds the currently registered destination access key.

`docs/migration-service.md` describes safety mechanisms that don't exist in the contract:

- Line 367 documents a `cancel_node_migration()` method ("removes the `OngoingNodeMigration` record… useful if the new node is not functioning correctly or wrong information was provided") — no such method exists.
- Lines 251 and 391 claim ongoing migrations are "automatically cleared when the protocol transitions from `Running` to `Resharing` or `Initializing`" — nothing clears `ongoing_migrations` on those transitions.

`docs/node-migration-guide.md` doesn't warn operators that starting a migration is effectively irreversible.

### Acceptance Criteria

- Either `cancel_node_migration()` is implemented (operator-callable, removes the caller's pending migration record), or the decision not to support cancellation is made explicit.
- `docs/migration-service.md` and `docs/node-migration-guide.md` accurately describe the implemented cancellation/expiry behavior.

### PR DESCRIPTION
**Pull request by metalurgical**
Implement cancel_node_migration() function in contract, which clears the ongoing migration record. 
Add tests.
Update migration-service documentation.

Related to #3774

### CODE REVIEW

File: crates/contract/src/lib.rs [hunk 1]
```diff
@@ -2845,6 +2845,32 @@ impl MpcContract {
...
+    ///
+    /// # Errors
+    /// - `NodeMigrationError::MigrationNotFound`: if no migration record exists for the caller
+    #[handle_result]
+    pub fn cancel_node_migration(&mut self) -> Result<(), Error> {
+        let account_id = Self::assert_caller_is_signer();
+
+        match self.node_migrations.remove_migration(&account_id) {
+            Some(destination_node_info) => log!(
+                "cancel_node_migration: signer={:?}, destination_node_info={:?}",
```
Comment (kevindeforth) [inline crates/contract/src/lib.rs]: The log can be emitted unconditionally

File: crates/contract/src/lib.rs [hunk 2]
```diff
@@ -2845,6 +2845,32 @@ impl MpcContract {
...
+    ///
+    /// Removes the caller's pending `DestinationNodeInfo` record. This is useful if the new node
+    /// is not functioning correctly or the wrong information was provided when calling
+    /// [`Self::start_node_migration`].
+    ///
+    /// This function is callable regardless of whether the protocol is in a `Running` state or
+    /// whether the signer is a current participant.
+    ///
+    /// # Errors
+    /// - `NodeMigrationError::MigrationNotFound`: if no migration record exists for the caller
```
Comment (kevindeforth) [inline crates/contract/src/lib.rs]: Please try to avoid paraphrasing the code in comments, c.f. [engineering-standards.md](https://github.com/near/mpc/blob/main/docs/engineering-standards.md)

File: crates/e2e-tests/tests/migration_service.rs [hunk 3]
```diff
@@ -996,3 +996,98 @@ async fn migration_service__should_handle_back_migration_a_to_b_to_a() {
         .await
         .expect("ckd request failed after back-migration");
 }
+
+/// Test to ensure `cancel_node_migration` removes a pending migration for the
+/// calling account.
```
Comment (kevindeforth) [inline crates/e2e-tests/tests/migration_service.rs]: idem

File: crates/contract/src/lib.rs [hunk 4]
```diff
@@ -8935,4 +8961,55 @@ mod tests {
...
+            .expect("expected at least one participant")
+            .clone();
+
+        testing_env!(
+            VMContextBuilder::new()
+                .signer_account_id(account_id.clone())
+                .predecessor_account_id(account_id.clone())
+                .attached_deposit(NearToken::from_yoctonear(1))
+                .build()
+        );
```
Comment (kevindeforth) [inline crates/contract/src/lib.rs]: you can use `Environment::new` for this https://github.com/near/mpc/blob/21d5ed34963ca6f7e9d226c1d4868d88e4fb2187/crates/contract/src/tee/test_utils.rs#L19

**Review by kevindeforth — COMMENTED** [review]
Thanks for the contribution!
Please avoid LLM-generated code-comments.
The unit-test in the contract can be improved by using existing test-helpers.

File: crates/contract/src/lib.rs (hunk 1 above)
Comment (metalurgical) [reply crates/contract/src/lib.rs]: Explain what you mean? destination_node_info is only available here if it was previously in the map

File: crates/contract/src/lib.rs (hunk 2 above)
Comment (metalurgical) [reply crates/contract/src/lib.rs]: Has been shortened considerably.

File: crates/e2e-tests/tests/migration_service.rs (hunk 3 above)
Comment (metalurgical) [reply crates/e2e-tests/tests/migration_service.rs]: Removed

File: crates/contract/src/lib.rs (hunk 4 above)
Comment (metalurgical) [reply crates/contract/src/lib.rs]: I see. Though it is done like the above quite extensively in the tests so it seemed like the correct way to do it here. Perhaps a separate issue should be logged to clean them up all at once?

**Comment by metalurgical** [comment]
Hmm, see refactors have merged ahead of this one, will update it in a short while.

**Comment by metalurgical** [comment]
@gilcu3 Updated

File: crates/contract/src/lib.rs [hunk 5]
```diff
@@ -7041,4 +7057,55 @@ mod tests {
...
     }
+
+    #[test]
+    fn test_cancel_node_migration() {
+        let running_state = ProtocolContractState::Running(gen_running_state(NUM_DOMAINS));
+        let mut contract = MpcContract::new_from_protocol_state(running_state);
+        let participants = {
+            let ProtocolContractState::Running(running) = &contract.protocol_state else {
+                panic!("expected running state");
+            };
```
Comment (gilcu3) [inline crates/contract/src/lib.rs]: nit: given this is just a test, we could simply use unwrap or expect here

File: crates/contract/src/lib.rs (hunk 4 above)
Comment (kevindeforth) [reply crates/contract/src/lib.rs]: Yeah, that's understandable, we are quite inconsistent in that regard.
It's not a blocker, so feel free to open an issue or leave as is.

File: crates/contract/src/lib.rs (hunk 1 above)
Comment (kevindeforth) [reply crates/contract/src/lib.rs]: A short log at the top, emitting predecessor-id is sufficient.
We don't need to log the removed destination node info.

File: crates/contract/src/lib.rs (hunk 2 above)
Comment (kevindeforth) [reply crates/contract/src/lib.rs]: Thank you, much better!

File: crates/contract/src/lib.rs (hunk 1 above)
Comment (metalurgical) [reply crates/contract/src/lib.rs]: Updated

File: crates/contract/src/lib.rs (hunk 4 above)
Comment (metalurgical) [reply crates/contract/src/lib.rs]: Ok, thanks will leave it as is for now.

**Comment by gilcu3** [comment]
@claude review

**Comment by gilcu3** [comment]
@metalurgical could you check the blocking points by Claude above? Once those are fixed I am ready to approve

File: crates/near-mpc-contract-interface/src/client.rs [hunk 6]
```diff
@@ -241,6 +241,17 @@ impl<C: CallContract> MpcContractHandle<C> {
         .await
     }
 
+    pub async fn cancel_node_migration(
```
Comment (kevindeforth) [inline crates/near-mpc-contract-interface/src/client.rs:252]: oh yeah, claude might have pointed it out already, but this should be added to the `mpc_contract_handle__should_match_the_wire_format_catalog` test

File: docs/migration-service.md [hunk 7]
```diff
@@ -321,7 +321,7 @@ flowchart TD
 For security reasons and to avoid edge cases and race conditions, the MPC network allows migration of nodes only while the protocol is in a `Running` state (as opposed to `Resharing` or `Initializing`, which are the two other well-defined states).
 
 Note that starting a migration workflow does not require a signing quorum. Instead, each participant can migrate their node at their own discretion. However, to avoid making the migration process a DoS attack vector, protocol state changes must have priority over any ongoing migrations.
-If the protocol state changes into a `Resharing` or `Initializing` state, any ongoing migration processes will simply be cancelled.
+If the protocol state changes into a `Resharing` or `Initializing` state, the pending `OngoingNodeMigration` record itself is **not** cleared by the transition and remains unless the operator explicitly withdraws it with `cancel_node_migration` or starts a new migration which will replace it.
```
Comment (kevindeforth) [inline docs/migration-service.md:324]: It wold be good to add a sentence that explains why this is ok:
a node concluding the migration sends its keyset to the contract, which includes the epoch id and all public keys. The contract compares that against the current epoch and keyset. The transaction will fail in case there is a discrepancy.

File: crates/contract/src/lib.rs (hunk 4 above)
Comment (metalurgical) [reply crates/contract/src/lib.rs]: Updated it now, while resolving conflicts for the latest refactor going in.

**Comment by metalurgical** [comment]
@gilcu3 The three blocking issues by claude should now be resolved. Have also resolved the conflicts (inclusive of moving this code to where the other bits went) and brought it back up to latest head again.

File: crates/near-mpc-contract-interface/src/client.rs (hunk 6 above)
Comment (metalurgical) [reply crates/near-mpc-contract-interface/src/client.rs:252]: Added

**Review by gilcu3 — COMMENTED** [review]
Thank you!

There is a small detail left, with a modified test failing. After that I can approve

**Comment by metalurgical** [comment]
@gilcu3 Thanks, passes now.

**Review by gilcu3 — DISMISSED** [review]

**Review by kevindeforth — DISMISSED** [review]
Thank you very much!

**Comment by gilcu3** [comment]
@metalurgical could you update main here?

**Comment by metalurgical** [comment]
> @metalurgical could you update main here?

Will check now

**Comment by metalurgical** [comment]
> > @metalurgical could you update main here?
> 
> Will check now

@gilcu3 

Looks like this:
```
self.contract
            .handle_for(self.operator_client_for(node_index)?)
```

Needed to become:
```
self.operator_client_for(node_index)?
            .call_mpc(self.contract_id())
```

Have committed the change. Just waiting for test to build and run locally.

**Review by gilcu3 — APPROVED** [review]

**Review by kevindeforth — APPROVED** [review]
